qiskit/transpiler/coupling.py

```diff
@@ -420,7 +420,7 @@ def largest_connected_component(self):
         """Return a set of qubits in the largest connected component."""
         return max(rx.weakly_connected_components(self.graph), key=len)
 
-    def connected_components(self) -> List["CouplingMap"]:
+    def connected_components(self, target=None) -> List["CouplingMap"]:
```

Comment:
This feels like a bit of a weird interface considering how we used things. We first create a `CouplingMap` from a `Target`, but then we have to pass the target to `connected_components` again when calling methods on the derived map?  It feels like something we perhaps ought to be encoding at construction time, maybe, or alternatively having `connected_components` to be an associated function that can be called either as a method or directly as a function passing a `target` in the first case?

Neither of those feel too great to me, though.

Reply:
Yeah, I wasn't happy with this interface really either. My first thought was that the target generated coupling map actually currently annotates the nodes and edges with some of the necessary information. But managing that was tricky and also the information wasn't exactly easy to use that way. I like your suggestion of using an associated method I can adjust the type signature on the disjoint functions to match what the passes are using to take in a target instead of a coupling map and then just generate it on the fly. It's not perfect but it at least avoids the back and forth between the objects.

Reply:
I reworked the filtering to be a function of the `Target` class and then updated the disjoint utils methods to build the coupling map connected components from either a target or a coupling map in: https://github.com/Qiskit/qiskit-terra/pull/9927/commits/c6a1e89d5211f9cd7cbcc70320272014a22cfcc7 I think this keeps the API separation a bit more clear as it's the target which owns all the aspects of which operations are supported on which qubits.